Implement per-request plugin state storage (#3286)

fspmarshall opened the pull request:
Implements a per-request storage API to allow plugins to store state with the auth server.  See #3286 for a full discussion.

This PR also addresses #3282 by adding retries for `CompareAndSwap` on `SetAccessRequestState` and and the new `UpdateAccessRequestPluginData` methods.

fspmarshall commented:
One part of this PR that I'm not completely happy with is the naming of `AccessRequestPluginDataUpdateParams`... In the original discussion in #3286 I used the much friendlier `PluginDataUpdateParams` name.  Given the size of the `services` namespace, I wasn't comfortable using such a generic name (after all, any future plugin systems will likely also end up using a `PluginData` abstraction of some kind).

I do have a solution to this dilemma, but I'm not certain if this is a case of me just over-engineering for a hypothetical future...

Basically, if we rename the `RequestID` field to `Resource` we can treat `PluginDataUpdateParams` as a generic construct which can be used for future plugin systems:

```golang
// current
AccessRequestPluginDataUpdateParams {
    RequestID: "...",
    Plugin: "...",
    //...
}

// proposed
PluginDataUpdateParams {
    Resource: "...",
    Plugin: "...",
   // ...
}
```

My intuition is to prefer this approach since it allows us to establish a standard by which per-resource plugin data is stored, across any future teleport plugin APIs.  I'm aware of my own tendency to focus too heavily on the general case, however, so I'm suspicious of this intuition.

@klizhentas any thoughts?

klizhentas commented:
`PluginDataUpdateParams` :+1:

russjones commented on lib/services/local/dynamic_access.go:
```diff
@@ -51,33 +52,40 @@ func (s *DynamicAccessService) CreateAccessRequest(ctx context.Context, req serv
-	if err != nil {
-		if trace.IsNotFound(err) {
-			return trace.NotFound("cannot set state of access request %q (not found)", name)
+	// Setting state is attempted multiple times in the event of concurrent writes.
```
Update comment to include example use-case.

russjones commented on lib/services/local/dynamic_access.go:
```diff
@@ -158,6 +171,134 @@ func (s *DynamicAccessService) UpsertAccessRequest(ctx context.Context, req serv
+	}
+	var matches []services.PluginData
+	for _, item := range result.Items {
+		if !bytes.HasSuffix(item.Key, []byte(paramsPrefix)) {
```
Comment why these items are being skipped.

russjones commented on lib/services/plugin_data.go:
```diff
@@ -0,0 +1,310 @@
+	return nil
+}
+
+func (d *PluginDataV3) checkExpectations(plugin string, expect map[string]string) error {
```
Comment missing.

klizhentas commented on lib/auth/auth_with_roles.go:
```diff
@@ -817,6 +817,33 @@ func (a *AuthWithRoles) SetAccessRequestState(ctx context.Context, reqID string,
 	return a.authServer.SetAccessRequestState(updateCtx, reqID, state)
 }
 
+func (a *AuthWithRoles) GetPluginData(ctx context.Context, filter services.PluginDataFilter) ([]services.PluginData, error) {
```
Missing godoc

klizhentas commented on lib/auth/clt.go:
```diff
@@ -2601,6 +2601,35 @@ func (c *Client) SetAccessRequestState(ctx context.Context, reqID string, state
+	if err != nil {
+		return nil, trail.FromGRPC(err)
+	}
+	data := make([]services.PluginData, 0, len(seq.PluginData))
```
What's the rationale behind the copy? it seems like sq.Data could be returned as is?

fspmarshall replied:
Required because we're casting the slice elements to `interface` values.

klizhentas commented on lib/auth/grpcserver.go:
```diff
@@ -246,6 +246,42 @@ func (g *GRPCServer) SetAccessRequestState(ctx context.Context, req *proto.Reque
 	return &empty.Empty{}, nil
 }
 
+func (g *GRPCServer) GetPluginData(ctx context.Context, filter *services.PluginDataFilter) (*proto.PluginDataSeq, error) {
```
missing godoc

klizhentas commented on lib/auth/grpcserver.go:
```diff
@@ -246,6 +246,42 @@ func (g *GRPCServer) SetAccessRequestState(ctx context.Context, req *proto.Reque
 }
 
+func (g *GRPCServer) GetPluginData(ctx context.Context, filter *services.PluginDataFilter) (*proto.PluginDataSeq, error) {
+	// TODO(fspmarshall): Add rate-limiting.
```
can you add more detail here?

klizhentas commented on lib/auth/grpcserver.go:
```diff
@@ -246,6 +246,42 @@ func (g *GRPCServer) SetAccessRequestState(ctx context.Context, req *proto.Reque
+}
+
+func (g *GRPCServer) UpdatePluginData(ctx context.Context, params *services.PluginDataUpdateParams) (*empty.Empty, error) {
+	// TODO(fspmarshall): Add rate-limiting.
```
can you add more data

klizhentas commented on lib/auth/grpcserver.go:
```diff
@@ -246,6 +246,42 @@ func (g *GRPCServer) SetAccessRequestState(ctx context.Context, req *proto.Reque
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	var seq []*services.PluginDataV3
```
why pointers here? You can use `nullable` modifier to prevent grpc from generating arrays of pointers, see examples:

https://github.com/gravitational/teleport/blob/master/lib/services/types.proto#L35

fspmarshall replied:
From what I can tell, this is consistent with how we handle `Resource` types elsewhere in the GRPC API.  At a guess, this is because we cast all our resource types to interfaces anyhow.

klizhentas commented on lib/auth/grpcserver.go:
```diff
@@ -246,6 +246,42 @@ func (g *GRPCServer) SetAccessRequestState(ctx context.Context, req *proto.Reque
+	}, nil
+}
+
+func (g *GRPCServer) UpdatePluginData(ctx context.Context, params *services.PluginDataUpdateParams) (*empty.Empty, error) {
```
missing godoc

klizhentas commented on lib/services/access_request_test.go:
```diff
@@ -51,6 +51,125 @@ func (s *AccessRequestSuite) TestRequestMarshaling(c *C) {
 	}
 }
 
+// TestPluginDataExpectations verifies the correct behavior of the `Expect` mapping.
```
what is a correct behavior, can you elaborate here?

klizhentas commented on lib/services/access_request_test.go:
```diff
@@ -51,6 +51,125 @@ func (s *AccessRequestSuite) TestRequestMarshaling(c *C) {
+			"spam":  "",
+		},
+	})
+	c.Assert(err, IsNil)
```
Should return `trace.CompareFailed` and use `fixtures.ExpectCompareFailed` to make sure the right error is returned

klizhentas commented on lib/services/local/dynamic_access.go:
```diff
@@ -51,33 +52,43 @@ func (s *DynamicAccessService) CreateAccessRequest(ctx context.Context, req serv
 }
 
 func (s *DynamicAccessService) SetAccessRequestState(ctx context.Context, name string, state services.RequestState) error {
```
missing doc

klizhentas commented on lib/services/local/dynamic_access.go:
```diff
@@ -51,33 +52,43 @@ func (s *DynamicAccessService) CreateAccessRequest(ctx context.Context, req serv
+	// The reason we bother to re-attempt is because state updates aren't meant
+	// to be "first come first serve".  Denials should overwrite approvals, but
+	// approvals should not overwrite denials.
+	for i := 0; i < maxCmpAttempts; i++ {
```
I haven't spotted any delay, randomization or backoff here? This will create a problem in production. Please use utils.Linear with some backoff

klizhentas commented on lib/services/local/dynamic_access.go:
```diff
@@ -101,6 +112,11 @@ func (s *DynamicAccessService) GetAccessRequests(ctx context.Context, filter ser
 		req, err := s.GetAccessRequest(ctx, filter.ID)
 		if err != nil {
+			if trace.IsNotFound(err) {
+				// Caller is going to need to check the length of the returned slice anyway,
```
I did not understand this comment. If something is not found, error should be returned, there should not be a scenario when we return nil, nil so to speak.

klizhentas commented on lib/services/local/dynamic_access.go:
```diff
@@ -158,6 +176,138 @@ func (s *DynamicAccessService) UpsertAccessRequest(ctx context.Context, req serv
+		if err != nil {
+			if trace.IsNotFound(err) {
+				// Caller is going to need to check the length of the returned slice anyway,
+				// so propagating a NotFoundError is redundant.
```
Disagree. So far in teleport and all other products, we never have 3 state problem:

* State 1: Object is correct, nil
* State 2: Object is incorrect, err
* State 3: nil, nil <- this third state is redundant, because every single caller now has to verify for thirds logical branch. 
If you only have State 1 and State 2, the code safely assumes, that if `err == nil` the object is valid. This part violates this invariant.

fspmarshall replied:
I initially agreed with this comment, and implemented appropriate changes, but that caused other errors which made me look closer... Turns out that we do in fact return `nil, nil` all the time!  We just don't generally do it explicitly.  Take, for example, this simplified version of the source of `GetUsers`:

```golang
result, _ := s.GetRange(...)
var out []services.User
for _, item := range result.Items {
	// do some conversions
	out = append(out, user)
}
return out, nil
```

If no users exist, `nil, nil` is in fact the return value of the function, and the rest of `teleport` handles this correctly.

The only reason why `nil, nil` is being returned *explicitly* in this PR is due to a performance optimization which short-circuits queries which match exactly zero or one resources.

klizhentas replied:
true, everything I said was meant for scalar values, for lists, returning a list of len 0 or nil is OK, is that this the case?

fspmarshall replied:
Yep, we're returning a slice here.

klizhentas commented on lib/services/access_request.go:
```diff
@@ -150,6 +150,10 @@ type DynamicAccess interface {
 	// DeleteAccessRequest deletes an access request.
 	DeleteAccessRequest(ctx context.Context, reqID string) error
+	// GetPluginData loads all plugin data matching the supplied filter.
+	GetPluginData(ctx context.Context, filter PluginDataFilter) ([]PluginData, error)
```
This method attempts to merge two functions: `GetPluginData` and `GetPluginDatas` it seems like, what is inconsistent with the rest of the API. Is there a reason for that?

fspmarshall replied:
This is consistent with how `GetAccessRequests` works.  There is no `GetAccessRequest` method; if you want to get a single resource, you just supply the resource name as part of the filter.

In terms of why this model originally arose, consistency with the new [watcher filters](https://github.com/gravitational/teleport/blob/798d2b552b74f9fec0f07cf0b9f47a01fd641a8a/lib/services/events.go#L51-L53) and optional parameters like `State` were a big factor.  It also played nicer with GRPC since a single message encapsulated all arguments.

Over the course of implementing this strategy for the `AccessRequest` resource it became obvious that creating a special method for getting a *single* access request would add a lot of additional code (new ephemeral GRPC messages in `auth.proto`, new methods for all of the different auth clients, new tests, and and so on) without any real benefit (a filter could already match a single resource).  I decided that a `GetAccessRequest` method was basically consistency hobgoblins at that point.

When it came time to implement the `PluginData` API, there were enough free parameters that the `Filter` strategy seemed appropriate here as well.  Like with the `AccessRequest` API, once a `Filter` based method exists, I just don't see the need for a single-resource getter.

klizhentas commented on lib/services/local/dynamic_access.go:
```diff
@@ -158,6 +176,138 @@ func (s *DynamicAccessService) UpsertAccessRequest(ctx context.Context, req serv
+	// Filters which specify Resource are a special case since they will match exactly zero or one
+	// possible PluginData instances.
+	if filter.Resource != "" {
+		item, err := s.Get(ctx, pluginDataKey(services.KindAccessRequest, filter.Resource))
```
Not sure I understand `filter` if filter is just a resource name, why not make it consistent with the rest of the API?

fspmarshall replied:
The `filter` also encodes resource kind and plugin name, it just isn't obvious reading this function body.  Resource kind is checked by the top-level `GetPluginData` function, and plugin name is used internally by the call to `filter.Match(...)`.

klizhentas commented on lib/services/local/dynamic_access.go:
```diff
@@ -158,6 +176,138 @@ func (s *DynamicAccessService) UpsertAccessRequest(ctx context.Context, req serv
+
+func (s *DynamicAccessService) updateAccessRequestPluginData(ctx context.Context, params services.PluginDataUpdateParams) error {
+	// Update is attempted multiple times in the event of concurrent writes.
+	for i := 0; i < maxCmpAttempts; i++ {
```
same here, with no backoff this will generate a spin lock with servers consuming 100% CPU on throttling or service outage.

klizhentas commented on lib/services/local/dynamic_access.go:
```diff
@@ -183,10 +333,47 @@ func itemToAccessRequest(item backend.Item) (services.AccessRequest, error) {
 }
 
+func itemFromPluginData(data services.PluginData) (backend.Item, error) {
+	const maxObjectSize = 1000000 // slightly less than 1mb
```
this constant should be in `teleport/constants.go`

klizhentas commented on lib/services/plugin_data.go:
```diff
@@ -0,0 +1,313 @@
+}
+
+func (f *PluginDataFilter) Match(data PluginData) bool {
+	if f.Kind != "" && f.Kind != data.GetSubKind() {
```
SubKind seems like a good candidate for plugin name?

fspmarshall replied:
`Subkind` is already used to encode the associated resource kind.  Currently it will always be `KindAccessRequest`, but using `SubKind` in this way ensures that we won't have any conflicts if we choose to expand the scope of `PluginData` later.

Even if `SubKind` were unused, its possible for more than one plugin to store data within a singe `PluginData` resource (hence the double-layer mapping in `PluginDataSpec`).

fspmarshall committed:
Implement per-resource PluginData storage (#3286)

- Also addresses #3282 by adding retries for CompareAndSwap
on SetAccessRequestState and UpdatePluginData.
